File: scripts/qc_fix_ld_matrix.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)

get_arg <- function(flag) {
  idx <- which(args == flag)
  if (length(idx) == 0 || idx == length(args)) {
    stop(sprintf("Missing required argument: %s", flag))
  }
  args[idx + 1]
}

input_matrix <- get_arg("--input-matrix")
out_matrix <- get_arg("--out-matrix")
diag_json <- get_arg("--diag-json")
done_file <- get_arg("--done-file")
tolerance <- as.numeric(get_arg("--tolerance"))

mkdir_for <- function(path) {
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
}

mkdir_for(out_matrix)
mkdir_for(diag_json)
mkdir_for(done_file)

mat_df <- read.table(input_matrix, sep = "\t", header = TRUE, check.names = FALSE, stringsAsFactors = FALSE)
if (ncol(mat_df) < 2) {
  stop("Input matrix file is malformed")
}

row_ids <- mat_df[[1]]
mat <- as.matrix(mat_df[, -1, drop = FALSE])
storage.mode(mat) <- "double"
col_ids <- colnames(mat)

if (nrow(mat) != ncol(mat)) {
  stop("LD matrix must be square")
}
if (length(row_ids) != nrow(mat) || !all(as.character(row_ids) == as.character(col_ids))) {
  stop("Row/column SNP IDs in matrix are not aligned")
}

before <- if (length(mat) == 0) 0.0 else max(abs(mat - t(mat)))
fixed <- (mat + t(mat)) / 2.0
if (nrow(fixed) > 0) {
  diag(fixed) <- 1.0
}
after <- if (length(fixed) == 0) 0.0 else max(abs(fixed - t(fixed)))

out <- data.frame(SNP = as.character(row_ids), fixed, check.names = FALSE)
colnames(out)[-1] <- as.character(row_ids)
write.table(out, file = out_matrix, sep = "\t", quote = FALSE, row.names = FALSE)

diag <- list(
  max_abs_asymmetry_before = as.numeric(before),
  max_abs_asymmetry_after = as.numeric(after),
  tolerance = as.numeric(tolerance),
  corrected = isTRUE(before > 0)
)

json_lines <- c(
  "{",
  sprintf("  \"max_abs_asymmetry_before\": %.12g,", diag$max_abs_asymmetry_before),
  sprintf("  \"max_abs_asymmetry_after\": %.12g,", diag$max_abs_asymmetry_after),
  sprintf("  \"tolerance\": %.12g,", diag$tolerance),
  sprintf("  \"corrected\": %s", tolower(as.character(diag$corrected))),
  "}"
)
writeLines(json_lines, diag_json)
writeLines("ok", done_file)

File: scripts/qc_fix_ld_matrix.py
#!/usr/bin/env python3
import argparse
import json
import os

import numpy as np
import pandas as pd


def _mkdir_for(path: str) -> None:
    os.makedirs(os.path.dirname(path), exist_ok=True)


def main() -> None:
    parser = argparse.ArgumentParser(description="QC and symmetrize LD matrix")
    parser.add_argument("--input-matrix", required=True)
    parser.add_argument("--out-matrix", required=True)
    parser.add_argument("--diag-json", required=True)
    parser.add_argument("--done-file", required=True)
    parser.add_argument("--tolerance", type=float, required=True)
    args = parser.parse_args()

    _mkdir_for(args.out_matrix)
    _mkdir_for(args.diag_json)
    _mkdir_for(args.done_file)

    mat_df = pd.read_csv(args.input_matrix, sep="\t", index_col=0)
    mat = mat_df.to_numpy(dtype=float)
    if mat.shape[0] != mat.shape[1]:
        raise ValueError("LD matrix must be square")

    before = float(np.max(np.abs(mat - mat.T))) if mat.size else 0.0
    fixed = (mat + mat.T) / 2.0
    np.fill_diagonal(fixed, 1.0)
    after = float(np.max(np.abs(fixed - fixed.T))) if fixed.size else 0.0

    if before > args.tolerance:
        # Keep going but record that correction was applied.
        pass

    out_df = pd.DataFrame(fixed, index=mat_df.index, columns=mat_df.columns)
    out_df.to_csv(args.out_matrix, sep="\t")

    diag = {
        "max_abs_asymmetry_before": before,
        "max_abs_asymmetry_after": after,
        "tolerance": args.tolerance,
        "corrected": bool(before > 0),
    }
    with open(args.diag_json, "w", encoding="utf-8") as handle:
        json.dump(diag, handle, indent=2)

    with open(args.done_file, "w", encoding="utf-8") as handle:
        handle.write("ok\n")


if __name__ == "__main__":
    main()
